src/main/scala/co/teapot/graph/TransposedGraphView.scala

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2016 Teapot, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
+ * file except in compliance with the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package co.teapot.graph
+
+/**
+  * A transposed view of a given graph.
+  */
+private[graph] class TransposedGraphView(graph: DirectedGraph) extends DirectedGraph {
+  override def outNeighbors(id: Int): IndexedSeq[Int] =
+    graph.inNeighbors(id)
+  override def inNeighbors(id: Int): IndexedSeq[Int] =
+    graph.outNeighbors(id)
+
+  override def outNeighbor(id: Int, i: Int): Int =
+    graph.inNeighbor(id, i)
+  override def inNeighbor(id: Int, i: Int): Int =
+    graph.outNeighbor(id, i)
+
+  override def outDegree(id: Int): Int = graph.inDegree(id)
+  override def inDegree(id: Int): Int = graph.outDegree(id)
+
+  // These just defer to the underlying graph.
+  override def maxNodeId: Int = graph.maxNodeId
```

Comment:
I think I now understand a little better what I was saying in the earlier comment. The earlier comment doesn't make sense, since you would have to define maxNodeId. But can we define an intermediate class, something like

class PackagedDirectedGraph(graph: DirectedGraph) extends Directed Graph

that defines all the functions as you have defined maxNodeId, eg. override def OutDegree(id: Int):Int = graph.inDegree(id)

And then, define TransPosedGraphView as extending PackagedDirectedGraph? That way, you have to only override the functions that you need to change.

It seems a little awkward to have to change every function in TransPosedGraphView and if we define additional views (eg. a BiDirectedView as in your other algorithms), this would make those simpler as well.

It is of course possible that there is an even simpler way of doing this which does not require going through PackagedDirectedGraph and I realize that the name is not ideal. But at least it requires us to only define the functionality that we change in TransPosedGraphView.